### PR TITLE
Cog environment variables

### PR DESCRIPTION
**Pull request by hangtwenty**
Implements #291.

This makes it so you can set environment variables in the Docker image that cog generates. In Dockerfile parlance, it's support for `ENV`.

This enables adding to cog.yaml:

```yaml
environment_variables:
    - FOOBAR=baz
    - SOME_LIBRARY_CACHE=/src/some_library_cache
```

And those become `ENV` directives in the Dockerfile.

----

It also applies a thoughtful default for [`$XDG_CACHE_HOME`](https://github.com/replicate/cog/issues/320#issuecomment-989545205) (see #320 for context). It's inside of `/src` so that it's preserved between runs.

### CODE REVIEW

File: pkg/dockerfile/generator.go [hunk 1]
```diff
@@ -131,9 +132,53 @@ func (g *Generator) baseImage() (string, error) {
...
+
+	if _, ok := envVarMap["XDG_CACHE_HOME"]; !ok {
+		// Cog sets a default value for $XDG_CACHE_HOME. Why:
+		// Cog mounts the project directory so anything in there (in /src in the
+		// image) gets retained between runs. $XDG_CACHE_HOME is used by various
+		// libraries including popular ML libraries; so, setting it to a subdirectory
+		// within /src results in retaining the cache between runs. Ultimately,
+		// everything in /src gets "baked in" to the cog image. So this is great
+		// for caching pretrained models and so on.
+		// For more context see: https://github.com/replicate/cog/issues/320
```
Comment (hangtwenty) [inline pkg/dockerfile/generator.go:165]: This comment is kinda verbose. I'd like to clean it up.

**Comment by bfirsh** [comment]
Great. Haven't reviewed this properly, but for starters: this should be `environment` for consistency with Docker Compose.

(Prepare for some noodling because this is my wheelhouse. 😁)

It's also worth noting that Docker Compose's `environment` is a _runtime_ feature rather than a _build_ feature. It is options passed to `docker run` rather than `ENV` statements in `Dockerfile`. It has the ability to read environment variables from the environment at runtime, etc.

I haven't thought this out in depth, but perhaps there is an alternate design here where environment variables feel like they're set at runtime. (But we do also want to set environment variables at build time to be baked into the image.)

It might be worth considering a future where we want to read environment variables from the environment. I wonder how that would work, and whether it semantically makes sense for them to live under `build.environment`. Where should options live in `cog.yaml` if they're used both at build time and run time?

Maybe this is a distinction that Cog doesn't need to make, because it is always in control of the image it is running, so perhaps `environment` makes sense as part of the build step. We can also move `environment` to somewhere else in the future if we change the semantics.

**Comment by hangtwenty** [comment]
@bfirsh yeah to be sure, the `environment_variables` name was verbatim from #291. I wasn't sure why `environment_variables` except that it might be to avoid confusion with the semantics of `environment` in Docker Compose.

**Comment by hangtwenty** [comment]
@bfirsh I grok what you mean about a design that considers runtime variables as well. This PR is meant to implement `ENV` directives for the generated `Dockerfile`s, while leaving it open for future features.

**Comment by hangtwenty** [comment]
In this first cut, `environment_variables` is under `build`.

I figure that the future features for runtime environment variables would live under `predict` rather than `build`. I say "under" because `predict` is currently top-level

(Notable precedent in `cog`'s current design: there's a `run` key under `build`, with semantics similar to `RUN` in Dockerfiles; i.e. it's run during build. So maybe instead of `build.environment_variables`, it's just `build.env`. `build.env`, `build.run` ... It makes sense to me anyways.)

**Comment by bfirsh** [comment]
To be clear — I might be wrong! Was just noodling on some possible design questions. environment in build is likely the right design…

**Pull request closed without merging**